src/shim/runtime/v2/rune/v2/create.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/containerd/containerd/runtime/v2/runc"
+	taskAPI "github.com/containerd/containerd/runtime/v2/task"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+var ContainerBase = "/run/enclave-cc/app"
+
+func create(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest) (*runc.Container, error) {
+	ociSpec, err := config.LoadSpec(filepath.Join(r.Bundle, configFilename))
+	if err != nil {
+		return nil, err
+	}
+
+	containerType, err := oci.ContainerType(*ociSpec)
+	if err != nil {
+		return nil, err
+	}
+	sandboxNamspace, err := oci.SandboxNamespace(*ociSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	var container *runc.Container
+
+	switch containerType {
+	case shimtypes.PodSandbox:
+		container, err = handlePodSandbox(ctx, s, r, sandboxNamspace)
+		if err != nil {
+			return nil, err
+		}
+	case shimtypes.PodContainer:
+		container, err = handlePodContainer(ctx, s, r, sandboxNamspace, ociSpec)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return container, nil
+}
+
+func handlePodSandbox(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest, sandboxNamspace string) (*runc.Container, error) {
+	container, err := runc.NewContainer(ctx, s.platform, r)
+	if err != nil {
+		return nil, err
+	}
+
+	if sandboxNamspace != shimtypes.KubeSystem {
+		ar := &taskAPI.CreateTaskRequest{
+			ID:       generateID(),
+			Terminal: false,
+			Options:  r.Options,
+		}
+
+		// Create agent enclave container
+		agentContainer, err := createAgentContainer(ctx, s, ar)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := writeAgentIdFile(r.Bundle, ar.ID); err != nil {
+			return nil, err
+		}
+
+		s.agentID = ar.ID
+		s.pauseID = r.ID
+		s.containers[ar.ID] = agentContainer
+		s.agent = &agent{
+			ID:     agentContainer.ID,
+			Bundle: agentContainer.Bundle,
+			URL:    AgentUrl,
+		}
+
+	}
+
+	return container, nil
+}
+
+func handlePodContainer(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest, sandboxNamspace string, ociSpec *specs.Spec) (*runc.Container, error) {
+	if sandboxNamspace != shimtypes.KubeSystem {
+		image, err := oci.GetImage(*ociSpec)
+		if err != nil {
+			return nil, err
+		}
+		cid, err := getContainerID(image)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create upperDir and workDir for app container
+		upperDir := filepath.Join(ContainerBase, r.ID, "upper")
+		workDir := filepath.Join(ContainerBase, r.ID, "work")
+		for _, dir := range []string{upperDir, workDir} {
+			if err := os.MkdirAll(dir, defaultDirPerm); err != nil && !os.IsExist(err) {
+				return nil, err
+			}
+		}
+		// sefsDir store the unionfs images (based on sefs)
+		sefsDir := filepath.Join(agentContainerRootDir, s.agentID, "merged/rootfs/images", cid)
+
+		var options []string
+		// Set index=off when mount overlayfs
+		options = append(options, "index=off")
+		options = append(options,
+			fmt.Sprintf("workdir=%s", workDir),
+			fmt.Sprintf("upperdir=%s", upperDir),
```

Comment:
Again, `[file]path.Join()` would be safer. It might be "impossible", but what if `r.ID == "foo %s bar"` for example. This function hasn't validated the request, so should trust it as little as possible.